### ISSUE TITLE
How to wrap a hook with a Provider with prop passing?

### ISSUE THREAD

**Issue by acrowell1-chwy** [testing-library/react-testing-library#1233]
### What is your question:

I have seen the following which was helpful: https://github.com/testing-library/react-hooks-testing-library/issues/23
This question steams further from that with regards to passing `initialProps`.  
I have updated React and DOM (`"react": "18.2.0", "react-dom": "18.2.0",`) and the React Testing Library (`"@testing-library/react": "14.0.0",`) to the most recent versions. I have followed that pattern, but notice that the props passed into Provider are only available via `children`. Please see the example below (borrowed from the tests within the library https://github.com/testing-library/react-hooks-testing-library/blob/1e01273374af4e48a0feb1f2233bf6c76d742167/src/__tests__/useContext.test.tsx): 

```Javascript 
 test('should update value in context when props are updated', () => {
    const TestContext = createContext('foo');

    const wrapper = ({
      current,
      children,
    }: {
      current: string;
      children: any;
    }) => {
      console.log({ current });
      console.log({ children });
      console.log('nested', children.props.renderCallbackProps);

      return (
        <TestContext.Provider value={current}>{children}</TestContext.Provider>
      );
    };

    const { result, rerender } = renderHook(() => useContext(TestContext), {
      wrapper,
      initialProps: {
        current: 'bar',
      },
    });

    rerender({ current: 'baz' });

    expect(result.current).toBe('baz');
  });

// console.logs

console.log
{ current: undefined }

console.log
{
  children: {
    '$$typeof': Symbol(react.element),
    type: [Function: TestComponent],
    key: null,
    ref: null,
    props: { renderCallbackProps: [Object] },
    _owner: null,
    _store: {}
  }
}

console.log
nested { current: 'baz' }


```

Was this the intended change? Are we supposed to only pass `children` to a `wrapper` and then desconstruct those props  from it? Can we update the tests or docs to represent the recommended pattern?

**Comment by mpeyper** [comment]
> @mpeyper you are correct, I am using `renderHook` from `@testing-library/react`. Yes, please move this to an issue. Thank you for confirming!

Done 🙂

FWIW, I believe their implementation should only pass children to the wrapper and never any props, so you will likely need to manage them manually outside of the `wrapper`/`renderHook` callbacks.

**Comment by acrowell1-chwy** [comment]
FWIW, I discovered that `props` passed from `initialProps` were nested within `children` -> `const { current } = children?.props?.renderCallbackProps;`. If that helps anyone. 
Appreciate the prompt response and aid. 😸 

Code example:  

```Typescript 
test("should update value in context when props are updated", () => {
  const TestContext = createContext("foo");

  const wrapper = ({ children }: { children: any }) => {
    const { current } = children?.props?.renderCallbackProps;
    return (
      <TestContext.Provider value={current}>{children}</TestContext.Provider>
    );
  };

  const { result, rerender } = renderHook(() => useContext(TestContext), {
    wrapper,
    initialProps: {
      current: "bar",
    },
  });

  rerender({ current: "baz" });

  expect(result.current).toBe("baz");
});

```

**Comment by acrowell1-chwy** [comment]
@mpeyper do you know if this `const { current } = children?.props?.renderCallbackProps;` is intentional for pulling props from `initialProps` (as shown above)?

**Comment by mpeyper** [comment]
I don’t know offhand and am not really involved with this implementation. My user standing was only the `TestComponent` was passed to the `wrapper` to be wrapped. Any props you are reading of the component is likely unsafe and not recommended.

**Comment by elboletaire** [comment]
This is how it is documented: https://react-hooks-testing-library.com/usage/advanced-hooks.

Considering it's not working as documented, this should be tagged as a bug, not as a question. Props should be available alongside the `children` one.

Anyone trying to test react 18 components following the documentation will face with this problem if they try to add any `initialProps`. Using the old `@testing-library/react-hooks` is not an option for anyone using react 18, since it throws errors about a react 18 deprecation in the render method.

Thanks for the workaround @acrowell1-chwy.

**Comment by mpeyper** [comment]
For other’s reference, the new docs are here https://testing-library.com/docs/react-testing-library/api/

**Comment by elboletaire** [comment]
> For other’s reference, the new docs are here [testing-library.com/docs/react-testing-library/api](https://testing-library.com/docs/react-testing-library/api/)

I guess you answered before my edits 🙃

**Comment by elboletaire** [comment]
> Whatever workarounds you use, please consider that it is not part of the supported API and could change without notice.

Yeah but the main API looks broken... the documented workaround (creating a `createWrapper` function) does not work for the `rerender` props update case. Shouldn't this be considered a bug then?

**Comment by davidgoli** [comment]
Agreed with the comments here, passing the "props" to the hook's render function (instead of the `wrapper` component) makes `rerender` much less useful. Perhaps I might recommend adding a second function like 'rerenderWrapper` that works this way? Something like:

```ts
const TrackingProvider = (props: { trackingProps: { userId: number }, children: React.ReactNode }) => {
  // .. Context provider that maintains state
}

// useTracking reads state & dispatches tracking events
const  { rerenderWrapper } = renderHook(() => useTracking(), {
  wrapper: TrackingProvider,
  initialProps:  {
    trackingProps: { userId: 123 }
  },
})

expect(track).toHaveBeenCalledWith({ userId: 123 })

rerender({ trackingProps: { userId: 456 })

expect(track).toHaveBeenCalledWith({ userId: 456 })
```

**Comment by admapop** [comment]
This is still an issue with the react-native flavour of the package.

**Comment by kkrivera-r7** [comment]
I've just hit this issue as well, especially around the rerender scenario. This worked much better in `@testing-library/react-hooks`
